File: msf-app/backend/app/services/db_connector.py
"""
Multi-Database Connector — mengambil metadata dari database yang sedang berjalan.
Support: PostgreSQL, MySQL, SQLite (MVP). SQL Server & MongoDB di v2.1.
"""

from typing import List, Optional
import re
from urllib.parse import quote_plus
from sqlalchemy import create_engine, text, inspect
from sqlalchemy.exc import OperationalError, SQLAlchemyError
import structlog

from app.models.schemas import (
    DBConnection, DBEngine,
    TableMetadata, ColumnMetadata, ForeignKeyMetadata, IndexMetadata,
    ViewMetadata, FunctionMetadata, DBMetadataResponse,
    DBTestConnectionResponse,
)

logger = structlog.get_logger()

# Default ports per engine
DEFAULT_PORTS = {
    DBEngine.POSTGRESQL: 5432,
    DBEngine.MYSQL: 3306,
    DBEngine.SQLITE: None,
    DBEngine.SQLSERVER: 1433,
    DBEngine.MONGODB: 27017,
}


class DBConnector:
    """
    Koneksi ke berbagai database engine dan ambil metadata.
    """

    @staticmethod
    def build_connection_url(conn: DBConnection) -> str:
        """Buat SQLAlchemy connection URL dari DBConnection model"""
        if conn.connection_string:
            return conn.connection_string

        engine = conn.engine
        port = conn.port or DEFAULT_PORTS.get(engine)
        username_enc = quote_plus(conn.username) if conn.username else ""
        password_enc = quote_plus(conn.password) if conn.password else ""

        if engine == DBEngine.POSTGRESQL or engine == "postgresql":
            return (
                f"postgresql+psycopg2://{username_enc}:{password_enc}"
                f"@{conn.host}:{port}/{conn.database}"
            )
        elif engine == DBEngine.MYSQL or engine == "mysql":
            return (
                f"mysql+pymysql://{username_enc}:{password_enc}"
                f"@{conn.host}:{port}/{conn.database}"
                f"?charset=utf8mb4"
            )
        elif engine == DBEngine.SQLITE or engine == "sqlite":
            # Untuk SQLite, database adalah path file
            return f"sqlite:///{conn.database}"
        elif engine == DBEngine.SQLSERVER or engine == "sqlserver":
            return (
                f"mssql+pyodbc://{username_enc}:{password_enc}"
                f"@{conn.host}:{port}/{conn.database}"
                f"?driver=ODBC+Driver+17+for+SQL+Server"
            )
        else:
            raise ValueError(f"Engine tidak didukung: {engine}")

    @classmethod
    async def test_connection(cls, conn: DBConnection) -> DBTestConnectionResponse:
        """Test apakah koneksi ke database berhasil (async non-blocking)"""
        import asyncio
        return await asyncio.to_thread(cls.test_connection_sync, conn)

    @classmethod
    def test_connection_sync(cls, conn: DBConnection) -> DBTestConnectionResponse:
        """Test apakah koneksi ke database berhasil (blocking sync)"""
        try:
            url = cls.build_connection_url(conn)
            connect_args = {}
            if conn.engine == DBEngine.POSTGRESQL or conn.engine == "postgresql":
                connect_args = {"connect_timeout": 10}
            elif conn.engine == DBEngine.MYSQL or conn.engine == "mysql":
                connect_args = {"connect_timeout": 10}
            elif conn.engine == DBEngine.SQLSERVER or conn.engine == "sqlserver":
                connect_args = {"timeout": 10}
            engine = create_engine(url, connect_args=connect_args)

            with engine.connect() as connection:
                # Cek versi server
                version_query = cls._get_version_query(conn.engine)
                result = connection.execute(text(version_query))
                version = result.scalar()

                # Dapatkan daftar schema & tabel per schema
                inspector = inspect(engine)
                try:
                    raw_schemas = inspector.get_schema_names()
                except Exception:
                    raw_schemas = []
                
                filtered_schemas = cls._filter_system_schemas(raw_schemas, conn.engine)
                
                tables_by_schema = {}
                for s in filtered_schemas:
                    try:
                        tables_by_schema[s] = inspector.get_table_names(schema=s)
                    except Exception:
                        try:
                            tables_by_schema[s] = inspector.get_table_names()
                        except Exception:
                            tables_by_schema[s] = []

                active_schema = conn.schema_name or ("main" if conn.engine == "sqlite" else "public")
                if active_schema not in tables_by_schema and filtered_schemas:
                    active_schema = filtered_schemas[0]
                
                table_count = len(tables_by_schema.get(active_schema, []))

            engine.dispose()

            return DBTestConnectionResponse(
                success=True,
                message="Koneksi berhasil",
                engine=conn.engine,
                server_version=str(version)[:100] if version else None,
                tables_count=table_count,
                schemas=filtered_schemas,
                tables_by_schema=tables_by_schema,
            )

        except OperationalError as e:
            logger.error("DB connection failed", error=str(e))
            return DBTestConnectionResponse(
                success=False,
                message=f"Gagal konek: {cls._clean_error(str(e))}",
                engine=conn.engine,
            )
        except Exception as e:
            logger.error("Unexpected DB error", error=str(e))
            return DBTestConnectionResponse(
                success=False,
                message=f"Error: {cls._clean_error(str(e))}",
                engine=conn.engine,
            )

    @classmethod
    def _filter_system_schemas(cls, schemas: List[str], engine: str) -> List[str]:
        """Menyaring schema bawaan sistem agar tidak muncul di pilihan"""
        system_schemas = {
            "postgresql": ["information_schema", "pg_catalog", "pg_toast", "pg_temp_1", "pg_toast_temp_1"],
            "mysql": ["information_schema", "performance_schema", "sys", "mysql"],
            "sqlite": ["main"],
        }
        filtered = []
        for s in schemas:
            s_lower = s.lower()
            if engine == "postgresql" and (s_lower.startswith("pg_") or s_lower in system_schemas["postgresql"]):
                continue
            if engine == "mysql" and s_lower in system_schemas["mysql"]:
                continue
            filtered.append(s)
        
        # Jika SQLite atau kosong, kembalikan schema default
        if not filtered:
            filtered = ["main"] if engine == "sqlite" else ["public"]
        return filtered

    @classmethod
    async def get_metadata(
        cls,
        conn: DBConnection,
        schema_filter: Optional[str] = None,
        table_filter: Optional[List[str]] = None,
        include_views: bool = False,
        include_functions: bool = False,
    ) -> DBMetadataResponse:
        """Ambil metadata lengkap dari database (async non-blocking)"""
        import asyncio
        return await asyncio.to_thread(
            cls.get_metadata_sync,
            conn,
            schema_filter,
            table_filter,
            include_views,
            include_functions,
        )

    @classmethod
    def get_metadata_sync(
        cls,
        conn: DBConnection,
        schema_filter: Optional[str] = None,
        table_filter: Optional[List[str]] = None,
        include_views: bool = False,
        include_functions: bool = False,
    ) -> DBMetadataResponse:
        """Ambil metadata lengkap dari database (blocking sync)"""
        url = cls.build_connection_url(conn)
        schema = schema_filter or conn.schema_name or "public"
        db_name = conn.database or "main"

        db_engine = create_engine(url)
        inspector = inspect(db_engine)

        try:
            tables = cls._get_tables(
                inspector, db_engine, conn.engine, schema, table_filter
            )
            views = cls._get_views(inspector, schema) if include_views else []
            functions = (
                cls._get_functions(db_engine, conn.engine, schema)
                if include_functions
                else []
            )

            return DBMetadataResponse(
                engine=conn.engine,
                database=db_name,
                schema=schema,
                tables=tables,
                views=views,
                functions=functions,
            )
        finally:
            db_engine.dispose()

    @classmethod
    def _get_tables(
        cls,
        inspector,
        db_engine,
        engine_type: str,
        schema: str,
        table_filter: Optional[List[str]] = None,
    ) -> List[TableMetadata]:
        """Ambil semua tabel beserta metadata kolom, PK, FK, index"""
        try:
            all_tables = inspector.get_table_names(schema=schema)
        except Exception:
            all_tables = inspector.get_table_names()

        if table_filter:
            all_tables = [t for t in all_tables if t in table_filter]

        tables = []
        for table_name in all_tables:
            try:
                table = cls._get_single_table(
                    inspector, db_engine, engine_type, table_name, schema
                )
                tables.append(table)
            except Exception as e:
                logger.warning(f"Gagal baca tabel {table_name}", error=str(e))

        return tables

    @classmethod
    def _get_single_table(
        cls, inspector, db_engine, engine_type: str, table_name: str, schema: str
    ) -> TableMetadata:
        """Ambil metadata satu tabel"""
        # Kolom
        try:
            raw_columns = inspector.get_columns(table_name, schema=schema)
        except Exception:
            raw_columns = inspector.get_columns(table_name)

        columns = []
        for col in raw_columns:
            columns.append(ColumnMetadata(
                name=col["name"],
                data_type=str(col["type"]),
                is_nullable=col.get("nullable", True),
                default_value=str(col["default"]) if col.get("default") is not None else None,
                is_primary_key=False,
                is_foreign_key=False,
            ))

        # Primary Key
        try:
            pk_info = inspector.get_pk_constraint(table_name, schema=schema)
            pk_cols = pk_info.get("constrained_columns", [])
        except Exception:
            pk_cols = []

        for col in columns:
            if col.name in pk_cols:
                col.is_primary_key = True

        # Foreign Keys
        try:
            raw_fks = inspector.get_foreign_keys(table_name, schema=schema)
        except Exception:
            raw_fks = []

        foreign_keys = []
        fk_col_names = set()
        for fk in raw_fks:
            for local_col, ref_col in zip(
                fk.get("constrained_columns", []),
                fk.get("referred_columns", [])
            ):
                fk_col_names.add(local_col)
                foreign_keys.append(ForeignKeyMetadata(
                    column=local_col,
                    references_table=fk.get("referred_table", ""),
                    references_column=ref_col,
                    on_delete=fk.get("options", {}).get("ondelete"),
                    on_update=fk.get("options", {}).get("onupdate"),
                ))

        for col in columns:
            if col.name in fk_col_names:
                col.is_foreign_key = True

        # Indexes
        try:
            raw_indexes = inspector.get_indexes(table_name, schema=schema)
        except Exception:
            raw_indexes = []

        indexes = []
        for idx in raw_indexes:
            # PostgreSQL functional indexes might return None in column_names
            cols = [c for c in idx.get("column_names", []) if c is not None]
            if not cols and idx.get("expressions"):
                cols = [str(expr) for expr in idx.get("expressions", []) if expr is not None]

            indexes.append(IndexMetadata(
                name=idx.get("name", f"idx_{table_name}") or f"idx_{table_name}",
                columns=cols,
                is_unique=idx.get("unique", False),
            ))

        # Row count (best effort)
        row_count = None
        try:
            with db_engine.connect() as db_conn:
                if engine_type == DBEngine.POSTGRESQL or engine_type == "postgresql":
                    # PostgreSQL estimate query
                    query = text("""
                        SELECT reltuples::bigint
                        FROM pg_class c
                        JOIN pg_namespace n ON n.oid = c.relnamespace
                        WHERE n.nspname = :schema AND c.relname = :table
                    """)
                    result = db_conn.execute(query, {"schema": schema, "table": table_name})
                    row_count = result.scalar()
                    if row_count is not None and row_count < 0:
                        row_count = 0
                elif engine_type == DBEngine.MYSQL or engine_type == "mysql":
                    # MySQL estimate query
                    query = text("""
                        SELECT TABLE_ROWS
                        FROM information_schema.tables
                        WHERE TABLE_SCHEMA = :schema AND TABLE_NAME = :table
                    """)
                    result = db_conn.execute(query, {"schema": schema, "table": table_name})
                    row_count = result.scalar()

                # Fallback to SELECT COUNT(*) if not PostgreSQL/MySQL or estimate failed (e.g. returns None)
                if row_count is None:
                    if engine_type == DBEngine.MYSQL or engine_type == "mysql":
                        escaped_table = table_name.replace('`', '``')
                        escaped_schema = schema.replace('`', '``')
                        quoted_table = f"`{escaped_schema}`.`{escaped_table}`"
                    elif engine_type == DBEngine.POSTGRESQL or engine_type == "postgresql":
                        escaped_table = table_name.replace('"', '""')
                        escaped_schema = schema.replace('"', '""')
                        quoted_table = f'"{escaped_schema}"."{escaped_table}"'
                    else: # SQLite
                        escaped_table = table_name.replace('"', '""')
                        quoted_table = f'"{escaped_table}"'
                    
                    result = db_conn.execute(text(f"SELECT COUNT(*) FROM {quoted_table}"))
                    row_count = result.scalar()
        except Exception:
            pass

        return TableMetadata(
            name=table_name,
            schema=schema,
            columns=columns,
            primary_key=pk_cols,
            foreign_keys=foreign_keys,
            indexes=indexes,
            row_count=row_count,
        )

    @classmethod
    def _get_views(cls, inspector, schema: str) -> List[ViewMetadata]:
        """Ambil list view"""
        views = []
        try:
            view_names = inspector.get_view_names(schema=schema)
            for name in view_names:
                try:
                    definition = inspector.get_view_definition(name, schema=schema)
                except Exception:
                    definition = None
                views.append(ViewMetadata(name=name, schema=schema, definition=definition))
        except Exception as e:
            logger.warning("Gagal ambil views", error=str(e))
        return views

    @classmethod
    def _get_functions(
        cls, db_engine, engine_type: str, schema: str
    ) -> List[FunctionMetadata]:
        """Ambil list stored functions/procedures (PostgreSQL only untuk MVP)"""
        functions = []
        if engine_type != DBEngine.POSTGRESQL:
            return functions

        query = text("""
            SELECT
                routine_name,
                routine_schema,
                external_language,
                data_type
            FROM information_schema.routines
            WHERE routine_schema = :schema
              AND routine_type = 'FUNCTION'
            ORDER BY routine_name
            LIMIT 100
        """)

        try:
            with db_engine.connect() as conn:
                result = conn.execute(query, {"schema": schema})
                for row in result:
                    functions.append(FunctionMetadata(
                        name=row[0],
                        schema=row[1],
                        language=row[2],
                        return_type=row[3],
                    ))
        except Exception as e:
            logger.warning("Gagal ambil functions", error=str(e))

        return functions

    @staticmethod
    def _get_version_query(engine_type: str) -> str:
        if engine_type == DBEngine.POSTGRESQL:
            return "SELECT version()"
        elif engine_type == DBEngine.MYSQL:
            return "SELECT VERSION()"
        elif engine_type == DBEngine.SQLITE:
            return "SELECT sqlite_version()"
        else:
            return "SELECT @@VERSION"

    @staticmethod
    def _get_table_count_query(engine_type: str, schema: str) -> str:
        if engine_type == DBEngine.POSTGRESQL:
            return f"SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = '{schema}' AND table_type = 'BASE TABLE'"
        elif engine_type == DBEngine.MYSQL:
            return "SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = DATABASE()"
        elif engine_type == DBEngine.SQLITE:
            return "SELECT COUNT(*) FROM sqlite_master WHERE type='table'"
        else:
            return "SELECT COUNT(*) FROM information_schema.tables"

    @staticmethod
    def _clean_error(error_msg: str) -> str:
        """Bersihkan pesan error dari info sensitif"""
        # Hilangkan password yang mungkin muncul di error message
        cleaned = re.sub(r"password=['\"]?[^'\"\s]+['\"]?", "password=***", error_msg, flags=re.IGNORECASE)
        # Batasi panjang
        return cleaned[:300]

File: msf-app/frontend/src/lib/server/dbConnector.ts
import { Client } from "pg";
import mysql from "mysql2/promise";
import { DBConnection, DBTestConnectionResponse, DBMetadataResponse, TableMetadata, ColumnMetadata, ForeignKeyMetadata, IndexMetadata } from "../types";

export class DBConnector {
  /**
   * Test database connection.
   */
  public static async testConnection(connection: DBConnection): Promise<DBTestConnectionResponse> {
    const engine = connection.engine;

    if (engine === "postgresql") {
      let client: Client | null = null;
      try {
        const config = this._getPgConfig(connection);
        client = new Client(config);
        await client.connect();

        // Query version and tables count
        const versionRes = await client.query("SELECT version()");
        const schema = connection.schema_name || "public";
        const tablesRes = await client.query(
          "SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = $1 AND table_type = 'BASE TABLE'",
          [schema]
        );

        const tablesCount = parseInt(tablesRes.rows[0].count, 10);
        return {
          success: true,
          message: "Koneksi ke PostgreSQL berhasil terhubung.",
          engine: "postgresql",
          server_version: versionRes.rows[0].version,
          tables_count: tablesCount,
        };
      } catch (err: any) {
        return {
          success: false,
          message: `Koneksi PostgreSQL gagal: ${this._cleanError(err.message || err)}`,
          engine: "postgresql",
        };
      } finally {
        if (client) await client.end().catch(() => {});
      }
    }

    if (engine === "mysql") {
      let conn: mysql.Connection | null = null;
      try {
        const config = this._getMysqlConfig(connection);
        conn = await mysql.createConnection(config);

        // Query version and tables count
        const [versionRows]: any = await conn.query("SELECT VERSION() as version");
        const [tablesRows]: any = await conn.query(
          "SELECT COUNT(*) as count FROM information_schema.tables WHERE table_schema = DATABASE() AND table_type = 'BASE TABLE'"
        );

        const serverVersion = versionRows[0]?.version || "Unknown";
        const tablesCount = parseInt(tablesRows[0]?.count || "0", 10);

        return {
          success: true,
          message: "Koneksi ke MySQL berhasil terhubung.",
          engine: "mysql",
          server_version: serverVersion,
          tables_count: tablesCount,
        };
      } catch (err: any) {
        return {
          success: false,
          message: `Koneksi MySQL gagal: ${this._cleanError(err.message || err)}`,
          engine: "mysql",
        };
      } finally {
        if (conn) await conn.end().catch(() => {});
      }
    }

    return {
      success: false,
      message: `Engine database '${engine}' tidak didukung di mode serverless. Silakan gunakan DDL SQL Paste.`,
      engine,
    };
  }

  /**
   * Fetch database metadata.
   */
  public static async fetchMetadata(
    connection: DBConnection,
    options: { schemaFilter?: string; includeViews?: boolean; includeFunctions?: boolean } = {}
  ): Promise<DBMetadataResponse> {
    const engine = connection.engine;

    if (engine === "postgresql") {
      let client: Client | null = null;
      try {
        const config = this._getPgConfig(connection);
        client = new Client(config);
        await client.connect();

        const schema = options.schemaFilter || connection.schema_name || "public";
        const tables = await this._inspectPostgres(client, schema);

        return {
          engine: "postgresql",
          database: connection.database || "postgres",
          schema,
          tables,
        };
      } catch (err: any) {
        throw new Error(`Gagal mengambil metadata PostgreSQL: ${this._cleanError(err.message || err)}`);
      } finally {
        if (client) await client.end().catch(() => {});
      }
    }

    if (engine === "mysql") {
      let conn: mysql.Connection | null = null;
      try {
        const config = this._getMysqlConfig(connection);
        conn = await mysql.createConnection(config);

        const tables = await this._inspectMysql(conn);

        return {
          engine: "mysql",
          database: connection.database || "mysql",
          schema: connection.database || "default",
          tables,
        };
      } catch (err: any) {
        throw new Error(`Gagal mengambil metadata MySQL: ${this._cleanError(err.message || err)}`);
      } finally {
        if (conn) await conn.end().catch(() => {});
      }
    }

    throw new Error(`Engine '${engine}' tidak didukung.`);
  }

  /**
   * Inspect PostgreSQL schema catalog
   */
  private static async _inspectPostgres(client: Client, schema: string): Promise<TableMetadata[]> {
    // 1. Get all base tables
    const tablesRes = await client.query(
      "SELECT table_name FROM information_schema.tables WHERE table_schema = $1 AND table_type = 'BASE TABLE' ORDER BY table_name",
      [schema]
    );

    const tables: TableMetadata[] = [];

    for (const row of tablesRes.rows) {
      const tableName = row.table_name;

      // 2. Fetch Columns
      const columnsRes = await client.query(
        `SELECT column_name, data_type, is_nullable, column_default, character_maximum_length 
         FROM information_schema.columns 
         WHERE table_schema = $1 AND table_name = $2 
         ORDER BY ordinal_position`,
        [schema, tableName]
      );

      // 3. Fetch Primary Keys
      const pkRes = await client.query(
        `SELECT kcu.column_name
         FROM information_schema.table_constraints tc
         JOIN information_schema.key_column_usage kcu
           ON tc.constraint_name = kcu.constraint_name
           AND tc.table_schema = kcu.table_schema
         WHERE tc.constraint_type = 'PRIMARY KEY'
           AND tc.table_schema = $1
           AND tc.table_name = $2`,
        [schema, tableName]
      );
      const pks = pkRes.rows.map((r: any) => r.column_name);

      // 4. Fetch Foreign Keys
      const fkRes = await client.query(
        `SELECT
             kcu.column_name AS column,
             ccu.table_name AS references_table,
             ccu.column_name AS references_column,
             rc.delete_rule AS on_delete,
             rc.update_rule AS on_update
         FROM information_schema.table_constraints tc
         JOIN information_schema.key_column_usage kcu
           ON tc.constraint_name = kcu.constraint_name
           AND tc.table_schema = kcu.table_schema
         JOIN information_schema.referential_constraints rc
           ON tc.constraint_name = rc.constraint_name
         JOIN information_schema.constraint_column_usage ccu
           ON rc.unique_constraint_name = ccu.constraint_name
           AND rc.unique_constraint_schema = ccu.table_schema
         WHERE tc.constraint_type = 'FOREIGN KEY'
           AND tc.table_schema = $1
           AND tc.table_name = $2`,
        [schema, tableName]
      );
      const fks: ForeignKeyMetadata[] = fkRes.rows.map((r: any) => ({
        column: r.column,
        references_table: r.references_table,
        references_column: r.references_column,
        on_delete: r.on_delete || undefined,
        on_update: r.on_update || undefined,
      }));

      // 5. Fetch Indexes
      const idxRes = await client.query(
        `SELECT
             ix.relname AS index_name,
             a.attname AS column_name,
             i.indisunique AS is_unique
         FROM pg_class t
         JOIN pg_index i ON t.oid = i.indrelid
         JOIN pg_class ix ON ix.oid = i.indexrelid
         JOIN pg_attribute a ON t.oid = a.attrelid AND a.attnum = ANY(i.indkey)
         JOIN pg_namespace n ON n.oid = t.relnamespace
         WHERE t.relkind = 'r'
           AND n.nspname = $1
           AND t.relname = $2`,
        [schema, tableName]
      );

      // Group index columns by index name
      const indexMap = new Map<string, { columns: string[]; isUnique: boolean }>();
      for (const idxRow of idxRes.rows) {
        const idxName = idxRow.index_name;
        const colName = idxRow.column_name;
        const isUnique = idxRow.is_unique;

        if (!indexMap.has(idxName)) {
          indexMap.set(idxName, { columns: [], isUnique });
        }
        indexMap.get(idxName)!.columns.push(colName);
      }

      const indexes: IndexMetadata[] = Array.from(indexMap.entries()).map(([name, val]) => ({
        name,
        columns: val.columns,
        is_unique: val.isUnique,
      }));

      // Compile columns metadata
      const columns: ColumnMetadata[] = columnsRes.rows.map((colRow: any) => {
        const colName = colRow.column_name;
        const isPk = pks.includes(colName);
        const isFk = fks.some(fk => fk.column === colName);

        return {
          name: colName,
          data_type: colRow.data_type.toUpperCase(),
          is_nullable: colRow.is_nullable === "YES",
          default_value: colRow.column_default || undefined,
          max_length: colRow.character_maximum_length || undefined,
          is_primary_key: isPk,
          is_foreign_key: isFk,
        };
      });

      // Get row count estimate
      const countRes = await client.query(
        `SELECT reltuples::bigint AS count FROM pg_class c
         JOIN pg_namespace n ON n.oid = c.relnamespace
         WHERE n.nspname = $1 AND c.relname = $2`,
        [schema, tableName]
      );
      const rowCount = countRes.rows[0] ? parseInt(countRes.rows[0].count, 10) : undefined;

      tables.push({
        name: tableName,
        schema,
        columns,
        primary_key: pks,
        foreign_keys: fks,
        indexes,
        row_count: rowCount !== undefined && rowCount >= 0 ? rowCount : undefined,
      });
    }

    return tables;
  }

  /**
   * Inspect MySQL schema catalog
   */
  private static async _inspectMysql(connection: mysql.Connection): Promise<TableMetadata[]> {
    // 1. Get all base tables
    const [tablesRows]: any = await connection.query(
      "SELECT table_name FROM information_schema.tables WHERE table_schema = DATABASE() AND table_type = 'BASE TABLE' ORDER BY table_name"
    );

    const tables: TableMetadata[] = [];

    for (const row of tablesRows) {
      const tableName = row.table_name || row.TABLE_NAME;

      // 2. Fetch Columns
      const [columnsRows]: any = await connection.query(
        `SELECT column_name, data_type, is_nullable, column_default, character_maximum_length, column_key
         FROM information_schema.columns 
         WHERE table_schema = DATABASE() AND table_name = ? 
         ORDER BY ordinal_position`,
        [tableName]
      );

      // 3. Fetch Foreign Keys
      const [fkRows]: any = await connection.query(
        `SELECT
             kcu.column_name AS \`column\`,
             kcu.referenced_table_name AS references_table,
             kcu.referenced_column_name AS references_column,
             rc.delete_rule AS on_delete,
             rc.update_rule AS on_update
         FROM information_schema.key_column_usage kcu
         JOIN information_schema.referential_constraints rc
           ON kcu.constraint_name = rc.constraint_name
           AND kcu.table_schema = rc.constraint_schema
         WHERE kcu.table_schema = DATABASE()
           AND kcu.table_name = ?
           AND kcu.referenced_table_name IS NOT NULL`,
        [tableName]
      );
      const fks: ForeignKeyMetadata[] = fkRows.map((r: any) => ({
        column: r.column || r.COLUMN,
        references_table: r.references_table || r.REFERENCES_TABLE,
        references_column: r.references_column || r.REFERENCES_COLUMN,
        on_delete: r.on_delete || undefined,
        on_update: r.on_update || undefined,
      }));

      // 4. Fetch Indexes
      const [idxRows]: any = await connection.query(
        `SELECT
             index_name,
             column_name,
             non_unique = 0 AS is_unique
         FROM information_schema.statistics
         WHERE table_schema = DATABASE()
           AND table_name = ?`,
        [tableName]
      );

      // Group indexes
      const indexMap = new Map<string, { columns: string[]; isUnique: boolean }>();
      for (const idxRow of idxRows) {
        const idxName = idxRow.index_name || idxRow.INDEX_NAME;
        const colName = idxRow.column_name || idxRow.COLUMN_NAME;
        const isUnique = parseInt(idxRow.is_unique || idxRow.IS_UNIQUE, 10) === 1;

        if (idxName === "PRIMARY") continue; // Handled separately via column key

        if (!indexMap.has(idxName)) {
          indexMap.set(idxName, { columns: [], isUnique });
        }
        indexMap.get(idxName)!.columns.push(colName);
      }

      const indexes: IndexMetadata[] = Array.from(indexMap.entries()).map(([name, val]) => ({
        name,
        columns: val.columns,
        is_unique: val.isUnique,
      }));

      // Compile columns & find Primary Keys
      const pks: string[] = [];
      const columns: ColumnMetadata[] = columnsRows.map((colRow: any) => {
        const colName = colRow.column_name || colRow.COLUMN_NAME;
        const colKey = colRow.column_key || colRow.COLUMN_KEY;
        const isPk = colKey === "PRI";
        const isFk = fks.some(fk => fk.column === colName);

        if (isPk) {
          pks.push(colName);
        }

        return {
          name: colName,
          data_type: (colRow.data_type || colRow.DATA_TYPE).toUpperCase(),
          is_nullable: (colRow.is_nullable || colRow.IS_NULLABLE) === "YES",
          default_value: colRow.column_default || colRow.COLUMN_DEFAULT || undefined,
          max_length: colRow.character_maximum_length || colRow.CHARACTER_MAXIMUM_LENGTH || undefined,
          is_primary_key: isPk,
          is_foreign_key: isFk,
        };
      });

      // Get row count estimate
      const [countRows]: any = await connection.query(
        "SELECT TABLE_ROWS as count FROM information_schema.tables WHERE table_schema = DATABASE() AND table_name = ?",
        [tableName]
      );
      const rowCount = countRows[0] ? parseInt(countRows[0].count, 10) : undefined;

      tables.push({
        name: tableName,
        schema: "default",
        columns,
        primary_key: pks,
        foreign_keys: fks,
        indexes,
        row_count: rowCount !== undefined && rowCount >= 0 ? rowCount : undefined,
      });
    }

    return tables;
  }

  /**
   * Get PG Connection configuration
   */
  private static _getPgConfig(connection: DBConnection): any {
    if (connection.connection_string) {
      return { connectionString: connection.connection_string };
    }
    return {
      host: connection.host || "localhost",
      port: connection.port || 5432,
      database: connection.database || "postgres",
      user: connection.username || "postgres",
      password: connection.password || "",
      ssl: connection.connection_string?.includes("sslmode=require") ? { rejectUnauthorized: false } : false,
    };
  }

  /**
   * Get MySQL Connection configuration
   */
  private static _getMysqlConfig(connection: DBConnection): any {
    if (connection.connection_string) {
      return connection.connection_string; // Uri connection string
    }
    return {
      host: connection.host || "localhost",
      port: connection.port || 3306,
      database: connection.database || "",
      user: connection.username || "root",
      password: connection.password || "",
    };
  }

  /**
   * Clean passwords from error messages
   */
  private static _cleanError(errStr: string): string {
    // Strips password parameters from standard URLs or connection strings
    return errStr.replace(/(:\/\/[\w_\-]+:)([^@]+)(@)/g, "$1******$3");
  }
}
